### PR TITLE
Multithreading in Python

### LINKED ISSUE #10: Multithreading in Python
## Basics
- What's Python Global Interpreter Lock

## Advanced

- Multithreading vs multiprocessing
- Process pool and thread pool executor
 - Python can only run I/O bound threads concurrently, CPU bound work is blocking (GIL)
- Alternatives to Multithreading:  
  - Asyncio
  - Coroutines
- Synchronisation primitives:
  - Channels
  - Locks
  - Mutexes

### PR DESCRIPTION
**Pull request by baffelli**
Closes #10

### CODE REVIEW

**Comment by despadam** [comment]
@baffelli Please double-check that my commit updates the skeleton of exercise 1 correctly. I used the code from the latest version of [the old branch](https://github.com/empa-scientific-it/python-tutorial/blob/new-material/parallelism/threads.ipynb). Without these changes, the notebook could not run.

**Review by despadam — CHANGES_REQUESTED** [review]
Thank you @baffelli for the great work! Most of the following are fixes for typos. However, as mentioned in my comments, I faced issues while running a couple of examples. I was also not able to pass the tests for the exercises. Would be great if you could double-check them, and then also show me what it is that perhaps I'm missing.

File: threads.ipynb [hunk 1]
```diff
@@ -0,0 +1,945 @@
...
+   "source": [
+    "%reload_ext tutorial.tests.testsuite"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Parallelism and concurrency in Python\n",
+    "\n",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb:42]: Done in c1bbcca

File: threads.ipynb [hunk 2]
```diff
@@ -0,0 +1,945 @@
...
+    "<figure>\n",
+    "  <img\n",
+    "  src=\"../../images/concurrency_vs_parallelism.jpg\"\n",
+    "  height=\"400px\"\n",
+    "  alt=\"The beautiful MDN logo.\">\n",
+    "  <figcaption>A simple time diagram illustrating the difference between parallelism and concurrency (source: https://openclassrooms.com/en/courses/5684021-scale-up-your-code-with-java-concurrency/5684028-identify-the-advantages-of-concurrency-and-parallelism)</figcaption>\n",
+    "</figure>\n",
+    "\n",
+    "\n",
+    "### Quiz: parallel or not\n",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb:89]: Done in 577747e

File: threads.ipynb [hunk 3]
```diff
@@ -0,0 +1,945 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We see that the first task to be started (`First task`) finished before the second one could start. \n",
+    "This is the sequential  computational model we are used to when we first learn programming. \n",
+    "However, in python we can introduce **parallelism** by using the [multiprocessing](https://docs.python.org/3/library/multiprocessing.html) module. \n",
+    "\n",
+    "Using this module, we can execute code in different operating system **processes**. \n",
+    "A process is a representation of a task, with all the code, memory and resources (files, network connections, etc.) needed to run it. \n",
+    "In most cases, processes are managed by the operating system, which takes care of scheduling what process should currently run and takes care that no process can run forever and regularly yields computing resources to other processes.\n",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb]: Changed in 148be2a

File: threads.ipynb [hunk 4]
```diff
@@ -0,0 +1,945 @@
...
+   "source": [
+    "### Threads, GIL and the illusion of concurrency\n",
+    "\n",
+    "Thus far, we only discussed one approach to concurrency (and parallelism) in python:  the use of operating system **processes** leveraging the ability of the OS to schedule and coordinate multiple processes across multiple CPUs. \n",
+    "\n",
+    "#### Threads vs processes\n",
+    "In reality, there's a second, very similar approach  called **threading**. \n",
+    "Like a **process**, a **thread** is a representation of a task including all needed resources. \n",
+    "In fact, a process usually consist of multiple threads all sharing a common memory. \n",
+    "Because of shared memory, a thread is usually lighter in its resource usage than a process, moreover the shared memory means that multiple threads can communicate using shared variables, although this style of concurrency comes with severe performance and safety pitfalls and it is being increasingly discouraged. \n",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb]: Changed in 148be2a

File: threads.ipynb [hunk 5]
```diff
@@ -0,0 +1,945 @@
...
+    "- It waits (joins) to the processes until they finish.\n",
+    "- It repeats the steps above until it exhausts the data to process\n",
+    "\n",
+    "Let's see how this works in practice with an artificial example: we call a function `work` that receives a number `i`, then waits 0.1 seconds and then return the number `i`. \n",
+    "We want to call this function `n` times and compute the sum of the results.\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb:353]: I don't get this problem. This is strange.

File: threads.ipynb [hunk 6]
```diff
@@ -0,0 +1,945 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Success! We can see that the sum computed in parallel using multiple processes equals the sum computed *locally* in one python process. Now it is interesting to try and see how much speedup we gain from this trick. To do so, we use the `timeit` module part of the python standard library:"
+   ]
+  },
+  {
+   "cell_type": "code",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb:398]: Same as above

File: threads.ipynb [hunk 7]
```diff
@@ -0,0 +1,945 @@
...
+    "When working on a multi-core or multi-CPU system, it is possible to leverage multitasking to run your computations in parallel. \n",
+    "We will see how in the following sections.\n",
+    "\n",
+    "### High-level interface: Process pools\n",
+    "\n",
+    "Let's rewrite our example from before using [`multiprocessing.Pool`](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.pool.Pool) which executes jobs on a pool of shared processes.  \n"
+   ]
+  },
+  {
+   "cell_type": "code",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb:185]: It worked for me.

File: threads.ipynb [hunk 8]
```diff
@@ -0,0 +1,945 @@
...
+    "Because a sum can computed in any order, we can have both producers work at the same time and get all the numbers and sum then in any order. \n",
+    "To exchange data between the producers and the consumers we use a `Queue`. \n",
+    "This is a list-like data structure where we put element into it from one side and we extract them from the other side; we call this *first in, first out* (FIFO). \n",
+    "The `multiprocessing` module offers an implementation of a queue that works with multiple processing, it is documented [here](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.Queue).\n",
+    "\n",
+    "Let's see how we do use a Queue with multiprocessing to solve the consumer-producer problem: "
+   ]
+  },
+  {
+   "cell_type": "code",
```
[the remark this answers is not in the document]
Comment (baffelli) [reply threads.ipynb:501]: I got the output myself. It is strange

**Comment by yakutovicha** [comment]
I set the PR "ready for review" since in fact it has been reviewed.

**Comment by baffelli** [comment]
> I set the PR "ready for review" since in fact it has been reviewed.

Wait, there are a couple of things that should be fixed still. Some of the problems with running the code that @despadam had should be investigated.

**Comment by yakutovicha** [comment]
> Wait, there are a couple of things that should be fixed still. Some of the problems with running the code that @despadam had should be investigated.

But "Ready for review" means only that it is ready for review, nothing more than that. We merge things only when they are approved.

**Comment by despadam** [comment]
Yes, just reviewing for now. Not merging yet, but it shouldn't be a draft anymore.

File: tutorial/threads.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+from .common import Question, Quiz
+
+
+class Threads(Quiz):
+    def __init__(self, title="Decide if the following are parallel or not"):
+        q1 = Question(
+            question="One cashier serves two lines of people in a store?",
+            options={
+                "Parallel:": "What if the cashier is slow?",
```
Comment (despadam) [inline tutorial/threads.py:9]:
```suggestion
                "Parallel": "What if the cashier is slow?",
```

**Review by despadam — CHANGES_REQUESTED** [review]
Just a small typo in the Quiz

File: tutorial/tests/test_threads.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+def find_word(letters: list[str], separator: str) -> bool:
+    """
+    This function finds a word in a list of letters separated by a separator.
+    """
+    return [w for w in "".join(letters).split(separator) if len(w) > 0]
+
+
+async def reference_exercise2(server: SecretServer) -> str:
+    """Write your solution here"""
+    rng = 30
```
[the remark this answers is not in the document]
Comment (baffelli) [reply tutorial/tests/test_threads.py]: I changed this to 50

**Comment by despadam** [comment]
For me the examples in the section **High-level interface: Process pools** cannot run. I get the following error in the console:

`
Process SpawnPoolWorker-1:
Traceback (most recent call last):
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\pool.py", line 114, in worker
    task = get()
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\queues.py", line 367, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'task' on <module '__main__' (built-in)>
Process SpawnPoolWorker-3:
Traceback (most recent call last):
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\process.py", line 314, in _bootstrap
    self.run()
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\pool.py", line 114, in worker
    task = get()
  File "C:\Users\adde\AppData\Local\miniconda3\envs\python-tutorial\lib\multiprocessing\queues.py", line 367, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'task' on <module '__main__' (built-in)>
`

**Review by despadam — DISMISSED** [review]

**Comment by despadam** [comment]
As agreed with @baffelli, it is best to merge and see if the examples and exercises will work on Binder, as this is the most pressing matter. It appears that the errors as OS dependent, so let's make sure that it will work online and then we can discourage the participants from running the tutorial locally, for this section at least.

**Comment by baffelli** [comment]
Let's try it [here](https://mybinder.org/v2/gh/empa-scientific-it/python-tutorial/new-material/parallelism-1)

**Comment by edoardob90** [comment]
Binder is unfortunately down now, and the error @despadam mentioned happens on my two machines as well. Not sure what we can do.

File: threads.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (edoardob90) [reply threads.ipynb:185]: Doesn't work for me either. Hangs indefinitely and cannot find the name `task`

File: threads.ipynb [hunk 11]
```diff
@@ -0,0 +1,1067 @@
...
+    "For example, we sit in a meeting, listen passively while working on our python program and stop working on our code to answer a question directed to us.\n",
+    "\n",
+    "\n",
+    "\n",
+    "The image below can help you understanding the difference between concurrent and parallel work.\n",
+    "<figure>\n",
+    "  <img\n",
+    "  src=\"../../images/concurrency_vs_parallelism.jpg\"\n",
+    "  height=\"400px\"\n",
+    "  alt=\"The beautiful MDN logo.\">\n",
```
Comment (edoardob90) [inline threads.ipynb:78]:
```suggestion
    "  alt=\"A schema of parallelism and concurrency\">\n",
```

File: threads.ipynb [hunk 12]
```diff
@@ -0,0 +1,1067 @@
...
+    "  <figcaption>A simple time diagram illustrating the difference between parallelism and concurrency (source: https://openclassrooms.com/en/courses/5684021-scale-up-your-code-with-java-concurrency/5684028-identify-the-advantages-of-concurrency-and-parallelism)</figcaption>\n",
```
Comment (edoardob90) [inline threads.ipynb:79]:
```suggestion
    "  <figcaption>A simple time diagram illustrating the difference between parallelism and concurrency (<a href=\"https://openclassrooms.com/en/courses/5684021-scale-up-your-code-with-java-concurrency/5684028-identify-the-advantages-of-concurrency-and-parallelism\">Source</a>)</figcaption>\n",
```

File: threads.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (yakutovicha) [reply threads.ipynb:185]: Doesn't work :(

File: threads.ipynb [hunk 13]
```diff
@@ -0,0 +1,1067 @@
...
+    "This is due to the fact that if we execute the function `work` sequentially for `n` inputs, we need to wait **at least** `n` times the duration of execution of `work`. \n",
+    "On the other hand, if we process in parallel, we can cut down the time by maximally `n_proc` where `n_proc` is the number of CPUs our system offers. \n",
+    "In reality, the speed up will be a bit smaller because of the overhead of starting different processes and the synchronization effort.\n",
+    "\n",
+    "\n",
+    "<div class=\"alert alert-block alert-warning\">\n",
+    "    <h4><b>Warning</b></h4> Multiprocessing/parallelism should be the last resort to improve the performance of your python code. \n",
+    "    You first should try improving your code by using the appropriate algorithms or by adopting numerical libraries like numpy that offer vectorized operations. \n",
+    "    If this fails, you could try just-in-time compilers like numba, which only require you to add a decorator to existing functions. Only if all these steps yield no improvement should you attempt to use multiprocessing explicitly. \n",
```
Comment (edoardob90) [inline threads.ipynb:486]:
```suggestion
    "    <h4><b>Warning</b></h4> Multiprocessing/parallelism should be the last resort to improve the performance of your Python code. \n",
    "    You first should try improving your code by using the appropriate algorithms or by adopting numerical libraries like NumPy that offers vectorized operations. \n",
    "    If this fails, you could try just-in-time compilers like <a href=\"https://numba.pydata.org\">Numba</a>, which only requires you to decorate existing functions. Only if all these steps yield no improvement, should you attempt to use multiprocessing explicitly. \n",
```

**Comment by baffelli** [comment]
> Binder is unfortunately down now, and the error @despadam mentioned happens on my two machines as well. Not sure what we can do.

That's really unfortunate. It seems to be highly OS dependent or maybe it's just a special case that runs on VSCode with the Jupiter extension. It might be that they implement a different Ipython kernel. I hope it works on Binder tomorrow, if not I'll try to walk people through examples in order to fill at least 1.5 hours

**Comment by edoardob90** [comment]
Minor changes to the text. Still, I **cannot** make the notebook work with any Python version I tested. I'm sure I'm missing something in the setup, but for me the entire notebook is in a "broken" state.

**Review by despadam — APPROVED** [review]
Fingers crossed

**Pull request merged**